pkg/scheduler/plugins/minruntime/minruntime.go

```diff
@@ -0,0 +1,212 @@
+	// store session data used for lookups
+	mr.queues = ssn.Queues
+	// setup caches on session open
+	mr.preemptProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+	mr.reclaimProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+
+	// initialize the resolver
+	mr.resolver = NewResolver(mr.queues, mr.defaultPreemptMinRuntime, mr.defaultReclaimMinRuntime)
+}
+
+func (mr *minruntimePlugin) OnSessionClose(ssn *framework.Session) {
+	mr.queues = nil
+	mr.preemptProtectionCache = nil
+	mr.reclaimProtectionCache = nil
+	mr.resolver = nil
+}
+
+func (mr *minruntimePlugin) reclaimFilterFn(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	protected := mr.isReclaimMinRuntimeProtected(pendingJob, victim)
+	// always return true for elastic jobs, but cache the result for scenario validator
+	if victim.IsElastic() {
+		return true
+	}
+
+	return !protected
+}
+
+func (mr *minruntimePlugin) preemptFilterFn(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	protected := mr.isPreemptMinRuntimeProtected(pendingJob, victim)
+	// always return true for elastic jobs, but cache the result for scenario validator
+	if victim.IsElastic() {
+		return true
+	}
+	return !protected
+}
+
+func (mr *minruntimePlugin) reclaimScenarioValidatorFn(scenario api.ScenarioInfo) bool {
+	reclaimer := scenario.GetPreemptor()
+	for _, victimInfo := range scenario.GetVictims() {
+		if !victimInfo.Job.IsElastic() {
+			continue
+		}
+		protected := mr.isReclaimMinRuntimeProtected(reclaimer, victimInfo.Job)
+		if !protected {
+			continue
+		}
+		numVictimTasks := int32(len(victimInfo.Tasks))
+		currentlyRunning := victimInfo.Job.GetActivelyRunningTasksCount()
+		if victimInfo.Job.MinAvailable > currentlyRunning-numVictimTasks {
+			return false
+		}
+	}
+	return true
+}
+
+func (mr *minruntimePlugin) preemptScenarioValidatorFn(scenario api.ScenarioInfo) bool {
+	preemptor := scenario.GetPreemptor()
+	for _, victimInfo := range scenario.GetVictims() {
+		if !victimInfo.Job.IsElastic() {
+			continue
+		}
+		protected := mr.isPreemptMinRuntimeProtected(preemptor, victimInfo.Job)
+		if !protected {
+			continue
+		}
+		numVictimTasks := int32(len(victimInfo.Tasks))
+		currentlyRunning := victimInfo.Job.GetActivelyRunningTasksCount()
+		if victimInfo.Job.MinAvailable > currentlyRunning-numVictimTasks {
+			return false
+		}
+	}
+	return true
+}
+
+func (mr *minruntimePlugin) isReclaimMinRuntimeProtected(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	if cached, ok := mr.reclaimProtectionCache[pendingJob.UID][victim.UID]; ok {
+		return cached
+	}
+	// Get the appropriate queue objects for both jobs
+	pendingQueue := mr.queues[pendingJob.Queue]
+	victimQueue := mr.queues[victim.Queue]
+
+	minRuntime, err := mr.resolver.getReclaimMinRuntime(mr.reclaimResolveMethod, pendingQueue, victimQueue)
```

Comment:
We should probably give some more context here - for example, reclaimer and reclaimee. Imagine if something goes wrong here due to some corruption in the session data (unlikely, admittedly) - you could default to a default min runtime, which could be 0 - meaning that the victim has no protection at all - this could have serious implications!
It's a very unlikely case, but let's be expressive as possible if it does happen. Maybe even resolve not allow the reclaim? (in cases of uncertainty, maybe it's better to "do no harm"?)

Reply:
I think minruntime should be seen as a best-effort protection rather than a hard guarantee. Because its arbitrarily set there is nothing that guarantees the job actually got to do its super-important initial work (it may have been delayed, slowed down by something that pushes the critical timeline outside of min-runtime), so the protection is best-effort as long as there is no scheduler<->workload communications to indicate readyness.
Ill add some more context, but I dont think we should refuse the reclaim.